File: drivers/zig/tests/url_test.zig
// URL parser fixtures. Mirrors the JS parser's edge cases so the
// same connection strings work everywhere.

const std = @import("std");
const reddb = @import("reddb");
const url = reddb.url;

const t = std.testing;

test "red://host" {
    var p = try url.parse(t.allocator, "red://example.com");
    defer p.deinit(t.allocator);
    try t.expectEqual(url.Kind.redwire, p.kind);
    try t.expectEqualStrings("example.com", p.host.?);
    try t.expectEqual(@as(u16, 5050), p.port.?);
}

test "red://host:port overrides default" {
    var p = try url.parse(t.allocator, "red://example.com:9999");
    defer p.deinit(t.allocator);
    try t.expectEqual(@as(u16, 9999), p.port.?);
}

test "reds:// flips TLS and keeps default port" {
    var p = try url.parse(t.allocator, "reds://example.com");
    defer p.deinit(t.allocator);
    try t.expectEqual(url.Kind.redwire_tls, p.kind);
    try t.expectEqual(@as(u16, 5050), p.port.?);
}

test "red:// with userinfo" {
    var p = try url.parse(t.allocator, "red://alice:s3cret@example.com");
    defer p.deinit(t.allocator);
    try t.expectEqualStrings("alice", p.username.?);
    try t.expectEqualStrings("s3cret", p.password.?);
    try t.expectEqualStrings("example.com", p.host.?);
}

test "red:// with token query" {
    var p = try url.parse(t.allocator, "red://h?token=sk-abc");
    defer p.deinit(t.allocator);
    try t.expectEqualStrings("sk-abc", p.token.?);
}

test "red:// with apiKey query" {
    var p = try url.parse(t.allocator, "red://h?apiKey=ak-xyz");
    defer p.deinit(t.allocator);
    try t.expectEqualStrings("ak-xyz", p.api_key.?);
}

test "red:// with api_key snake-case" {
    var p = try url.parse(t.allocator, "red://h?api_key=ak-xyz");
    defer p.deinit(t.allocator);
    try t.expectEqualStrings("ak-xyz", p.api_key.?);
}

test "red:// with loginUrl override" {
    var p = try url.parse(t.allocator, "red://h?loginUrl=https%3A%2F%2Faux%2Fauth%2Flogin");
    defer p.deinit(t.allocator);
    try t.expectEqualStrings("https://aux/auth/login", p.login_url.?);
}

test "red:// proto=https switches kind" {
    var p = try url.parse(t.allocator, "red://h?proto=https");
    defer p.deinit(t.allocator);
    try t.expectEqual(url.Kind.https, p.kind);
}

test "red:// proto=pg picks default port 5432" {
    var p = try url.parse(t.allocator, "red://h?proto=pg");
    defer p.deinit(t.allocator);
    try t.expectEqual(url.Kind.pg, p.kind);
}

test "red:// proto=grpcs uses 5052 default" {
    var p = try url.parse(t.allocator, "red://h?proto=grpcs");
    defer p.deinit(t.allocator);
    try t.expectEqual(url.Kind.grpcs, p.kind);
}

test "red:// proto=unknown → error" {
    try t.expectError(reddb.Error.UnsupportedProto, url.parse(t.allocator, "red://h?proto=ftp"));
}

test "red:/// embedded path" {
    var p = try url.parse(t.allocator, "red:///var/lib/red.rdb");
    defer p.deinit(t.allocator);
    try t.expectEqual(url.Kind.embedded, p.kind);
    try t.expectEqualStrings("/var/lib/red.rdb", p.path.?);
}

test "red:// alone is embedded" {
    var p = try url.parse(t.allocator, "red://");
    defer p.deinit(t.allocator);
    try t.expectEqual(url.Kind.embedded, p.kind);
}

test "red://memory is embedded" {
    var p = try url.parse(t.allocator, "red://memory");
    defer p.deinit(t.allocator);
    try t.expectEqual(url.Kind.embedded, p.kind);
}

test "memory:// legacy form" {
    var p = try url.parse(t.allocator, "memory://");
    defer p.deinit(t.allocator);
    try t.expectEqual(url.Kind.embedded, p.kind);
}

test "file:// legacy with path" {
    var p = try url.parse(t.allocator, "file:///tmp/x.rdb");
    defer p.deinit(t.allocator);
    try t.expectEqual(url.Kind.embedded, p.kind);
    try t.expectEqualStrings("/tmp/x.rdb", p.path.?);
}

test "grpc:// legacy parses" {
    var p = try url.parse(t.allocator, "grpc://h:5051");
    defer p.deinit(t.allocator);
    try t.expectEqual(url.Kind.grpc, p.kind);
    try t.expectEqual(@as(u16, 5051), p.port.?);
}

test "grpcs:// legacy parses" {
    var p = try url.parse(t.allocator, "grpcs://h");
    defer p.deinit(t.allocator);
    try t.expectEqual(url.Kind.grpcs, p.kind);
    try t.expectEqual(@as(u16, 5052), p.port.?);
}

test "http:// parses host and port" {
    var p = try url.parse(t.allocator, "http://h:8080");
    defer p.deinit(t.allocator);
    try t.expectEqual(url.Kind.http, p.kind);
    try t.expectEqual(@as(u16, 8080), p.port.?);
}

test "https:// default port 8443" {
    var p = try url.parse(t.allocator, "https://h");
    defer p.deinit(t.allocator);
    try t.expectEqual(url.Kind.https, p.kind);
    try t.expectEqual(@as(u16, 8443), p.port.?);
}

test "unknown scheme errors" {
    try t.expectError(reddb.Error.UnsupportedScheme, url.parse(t.allocator, "ftp://x"));
}

test "empty uri errors" {
    try t.expectError(reddb.Error.UnparseableUri, url.parse(t.allocator, ""));
}

test "red:// percent-encoded password" {
    var p = try url.parse(t.allocator, "red://alice:p%40ss@h");
    defer p.deinit(t.allocator);
    try t.expectEqualStrings("p@ss", p.password.?);
}

File: drivers/cpp/tests/url_test.cpp
// URL parser tests. Targets ≥20 cases across schemes, ports, auth,
// query strings and embedded rejection.

#include "reddb/errors.hpp"
#include "reddb/url.hpp"

#include <gtest/gtest.h>

using namespace reddb;

TEST(Url, RedDefaultPort) {
    auto u = parse_uri("red://localhost");
    EXPECT_EQ(u.kind, UrlKind::Red);
    EXPECT_EQ(u.host, "localhost");
    EXPECT_EQ(u.port, 5050);
}

TEST(Url, RedExplicitPort) {
    auto u = parse_uri("red://example.com:6000");
    EXPECT_EQ(u.kind, UrlKind::Red);
    EXPECT_EQ(u.port, 6000);
}

TEST(Url, RedsExplicitTls) {
    auto u = parse_uri("reds://primary:5051");
    EXPECT_EQ(u.kind, UrlKind::Reds);
    EXPECT_EQ(u.port, 5051);
}

TEST(Url, RedsDefaultPort) {
    auto u = parse_uri("reds://primary");
    EXPECT_EQ(u.kind, UrlKind::Reds);
    EXPECT_EQ(u.port, 5051);
}

TEST(Url, HttpDefaultPort) {
    auto u = parse_uri("http://api.example.com");
    EXPECT_EQ(u.kind, UrlKind::Http);
    EXPECT_EQ(u.port, 8080);
}

TEST(Url, HttpsDefaultPort) {
    auto u = parse_uri("https://api.example.com");
    EXPECT_EQ(u.kind, UrlKind::Https);
    EXPECT_EQ(u.port, 8443);
}

TEST(Url, HttpExplicitPort) {
    auto u = parse_uri("https://api.example.com:9443");
    EXPECT_EQ(u.kind, UrlKind::Https);
    EXPECT_EQ(u.port, 9443);
}

TEST(Url, UserPassFromAuthority) {
    auto u = parse_uri("red://alice:hunter2@db.local:5050");
    ASSERT_TRUE(u.username);
    ASSERT_TRUE(u.password);
    EXPECT_EQ(*u.username, "alice");
    EXPECT_EQ(*u.password, "hunter2");
}

TEST(Url, UserOnlyFromAuthority) {
    auto u = parse_uri("red://alice@db.local");
    ASSERT_TRUE(u.username);
    EXPECT_FALSE(u.password.has_value());
    EXPECT_EQ(*u.username, "alice");
}

TEST(Url, PercentDecodedUserInfo) {
    auto u = parse_uri("red://al%40ice:p%40ss@db.local");
    EXPECT_EQ(*u.username, "al@ice");
    EXPECT_EQ(*u.password, "p@ss");
}

TEST(Url, TokenQueryParam) {
    auto u = parse_uri("red://db.local:5050?token=sk-abc");
    ASSERT_TRUE(u.token);
    EXPECT_EQ(*u.token, "sk-abc");
}

TEST(Url, ApiKeyAlias) {
    auto u1 = parse_uri("red://db.local?apiKey=ak-1");
    auto u2 = parse_uri("red://db.local?api_key=ak-2");
    ASSERT_TRUE(u1.api_key);
    ASSERT_TRUE(u2.api_key);
    EXPECT_EQ(*u1.api_key, "ak-1");
    EXPECT_EQ(*u2.api_key, "ak-2");
}

TEST(Url, LoginUrlAlias) {
    auto u1 = parse_uri("red://db.local?loginUrl=https://x/login");
    auto u2 = parse_uri("red://db.local?login_url=https://y/login");
    ASSERT_TRUE(u1.login_url);
    ASSERT_TRUE(u2.login_url);
    EXPECT_EQ(*u1.login_url, "https://x/login");
    EXPECT_EQ(*u2.login_url, "https://y/login");
}

TEST(Url, ProtoOverrideHttps) {
    auto u = parse_uri("red://db.local:8443?proto=https");
    EXPECT_EQ(u.kind, UrlKind::Https);
}

TEST(Url, ProtoOverrideReds) {
    auto u = parse_uri("red://db.local?proto=reds");
    EXPECT_EQ(u.kind, UrlKind::Reds);
}

TEST(Url, ProtoOverrideGrpcsBecomesReds) {
    auto u = parse_uri("red://db.local?proto=grpcs");
    EXPECT_EQ(u.kind, UrlKind::Reds);
}

TEST(Url, EmptyUriRejected) {
    EXPECT_THROW(parse_uri(""), RedDBError);
}

TEST(Url, EmbeddedRedRejected) {
    EXPECT_THROW(parse_uri("red://"), RedDBError);
}

TEST(Url, EmbeddedRedPathRejected) {
    EXPECT_THROW(parse_uri("red:///var/lib/data.rdb"), RedDBError);
}

TEST(Url, EmbeddedMemoryRejected) {
    EXPECT_THROW(parse_uri("memory://"), RedDBError);
}

TEST(Url, EmbeddedFileRejected) {
    EXPECT_THROW(parse_uri("file:///tmp/x.rdb"), RedDBError);
}

TEST(Url, UnsupportedSchemeRejected) {
    EXPECT_THROW(parse_uri("mongodb://localhost"), RedDBError);
}

TEST(Url, BadProtoRejected) {
    EXPECT_THROW(parse_uri("red://db.local?proto=ftp"), RedDBError);
}

TEST(Url, OriginalUriPreserved) {
    auto u = parse_uri("red://localhost:5050?token=abc");
    EXPECT_EQ(u.original_uri, "red://localhost:5050?token=abc");
}

TEST(Url, GrpcAliasMapsToRed) {
    auto u = parse_uri("grpc://localhost:5051");
    EXPECT_EQ(u.kind, UrlKind::Red);
    EXPECT_EQ(u.port, 5051);
}

TEST(Url, GrpcsAliasMapsToReds) {
    auto u = parse_uri("grpcs://localhost:5052");
    EXPECT_EQ(u.kind, UrlKind::Reds);
    EXPECT_EQ(u.port, 5052);
}

TEST(Url, ErrorCodeForEmbedded) {
    try {
        parse_uri("red:///abs/path");
        FAIL() << "expected EmbeddedUnsupported";
    } catch (const RedDBError& e) {
        EXPECT_EQ(e.code(), ErrorCode::EmbeddedUnsupported);
    }
}

TEST(Url, BadPortRejected) {
    EXPECT_THROW(parse_uri("red://localhost:abc"), RedDBError);
}

TEST(Url, MissingHostRejected) {
    // `red://?proto=https` — no host. raw_parse picks up empty
    // host and we throw InvalidUri.
    EXPECT_THROW(parse_uri("red://?proto=https"), RedDBError);
}
